### PR TITLE
NetworkPolicy - OSDOCS 374

### PR DESCRIPTION
**Pull request by jboxman**
- https://jira.coreos.com/browse/OSDOCS-374

### CODE REVIEW

**Comment by jboxman** [comment]
@openshift/team-documentation, as a sanity check when converting this from 3.x, do each of the bulleted items become procedure modules and the preamble a concept module?

Thanks.

**Comment by kalexand-rh** [comment]
@jboxman, yes the intro should be a concept. I'm not sure which user stories you're supporting, so I can't say if you'd have one procedure or a procedure per li.

**Comment by jboxman** [comment]
@openshift/team-documentation, PTAL. Thanks.

**Comment by jboxman** [comment]
@squeed @danwinship @pecameron

The network policy content in this PR is migrated from 3.x. What changes are needed for 4.x? Thanks.

File: modules/nw-networkpolicy-about.adoc [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// Module included in the following assemblies:
+//
+// networking/configuring-networkpolicy.adoc
+
+[id="nw-networkpolicy-about-{context}"]
+
+= About network policy
+
+In {product-title} {product-version} `NetworkPolicy` is the default mode for the
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Suggest a comma after {product-version}

File: modules/nw-networkpolicy-about.adoc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+`OpenShiftSDN` plug-in for the Kubernetes Container Network Interface (CNI).
+
+[NOTE]
+====
+The `v1` NetworkPolicy features are available only in {product-title}. This
+means that egress policy types, IPBlock, and combining `podSelector` and
+`namespaceSelector` are not available in {product-title}.
+====
+
+In a cluster network isolation is controlled entirely by
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Suggest a comma after "In a cluster"

File: modules/nw-networkpolicy-about.adoc (hunk 1 above)
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Also wondering whether "NetworkPolicy" really needs to be in backticks everywhere. For things like ConfigMap, DeploymentConfig, etc., the guidance is to not use backticks.

File: modules/nw-networkpolicy-about.adoc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/network-policy.md[`NetworkPolicy`
+objects]. By default, all pods in a project are accessible from other pods and
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: All instances of pod/pods lowecase should be capitalized Pod/Pods - throughout all files in the PR.

File: modules/nw-networkpolicy-about.adoc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+`NetworkPolicy` objects in that project to indicate the allowed incoming
+connections. Project administrators can create and delete `NetworkPolicy`
+objects within their own project.
+
+Pods that do not have `NetworkPolicy` objects pointing to them are fully
+accessible, whereas, pods that have one or more `NetworkPolicy` objects pointing
+to them are isolated. These isolated pods only accept connections that are
+accepted by at least one of their `NetworkPolicy` objects.
+
+Following are a few sample `NetworkPolicy` object definitions supporting
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Suggest rewording to not start the sentence with just "Following". Maybe something like "The following example NetworkPolicy objects demonstrate supporting different scenarios."

File: modules/nw-networkpolicy-about.adoc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+different scenarios:
+
+* *Deny All Traffic*
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: I don't think we use asterisks except for referring to GUI items. Suggest instead trying this syntax:

```
Deny all traffic::
```

Also, these should all be sentence case.

File: modules/nw-networkpolicy-about.adoc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
++
+To make a project "deny by default" add a `NetworkPolicy` object that
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Don't think the quotes are necessary here.

File: modules/nw-networkpolicy-about.adoc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  name: deny-by-default
+spec:
+  podSelector:
+  ingress: []
+----
+
+* *Only Accept connections from pods within project*
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: withing project -> within (a or the) project

File: modules/nw-networkpolicy-create.adoc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+* A cluster using the OpenShiftSDN network plug-in with `mode: NetworkPolicy`
+set.
+* The `oc` CLI program must be installed.
+* You must be logged in to the cluster and your user must have the `admin` role
+for the project you are creating policy objects in.
+
+.Procedure
+
+. Create a policy rule:
+** Create a `<policy-name>.yaml` file where `<policy-name>` describes the policy
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-create.adoc]: Suggest using ".." instead of "**" because these are still ordered substeps.

File: modules/nw-networkpolicy-create.adoc [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+rule.
+** In the file you just created define a policy object, such as in the following
+example:
++
+----
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-create.adoc]: Add [source,yaml] above this

File: modules/nw-networkpolicy-create.adoc [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+. Run the following command to create the policy object:
++
+----
+$ oc create -f <policy-name>.yaml -n <project>
+----
++
+In the following example, a new `NetworkPolicy` object is created in a project
+named `project1`:
+----
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-create.adoc]: Need a "+" in the line above this, so that the following code is indented properly with step 2.

File: modules/nw-networkpolicy-delete.adoc [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+* A cluster using the OpenShiftSDN network plug-in with `mode: NetworkPolicy`
+set.
+* The `oc` CLI program must be installed.
+* You must be logged in to the cluster and your user must have the `admin` role
+for the project you are creating policy objects in.
+
+.Procedure
+
+. To delete a `NetworkPolicy` object, run the following command:
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-delete.adoc]: When there is only 1 step in a procedure, use a "*" so it's unordered.

File: modules/nw-networkpolicy-view.adoc [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+.Prerequisites
+
+* A cluster using the OpenShiftSDN network plug-in with `mode: NetworkPolicy`
+set.
+* The `oc` CLI program must be installed.
+* You must be logged in to the cluster.
+
+.Procedure
+
+. To view `NetworkPolicy` objects defined in your cluster, run the following
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-view.adoc]: Same about using * for only 1 step

File: modules/nw-networkpolicy-about.adoc [hunk 13]
```diff
@@ -0,0 +1,100 @@
+// Module included in the following assemblies:
+//
+// networking/configuring-networkpolicy.adoc
+
+[id="nw-networkpolicy-about-{context}"]
+
+= About network policy
+
+In {product-title} {product-version} `NetworkPolicy` is the default mode for the
+`OpenShiftSDN` plug-in for the Kubernetes Container Network Interface (CNI).
```
Comment (bergerhoffer) [inline modules/nw-networkpolicy-about.adoc]: Same question on if this needs to be in backticks. The other places in this PR don't, and I don't think it's necessary.

File: networking/configuring-networkpolicy.adoc [hunk 14]
```diff
@@ -0,0 +1,16 @@
+[id="configuring-networkpolicy-plugin"]
+= Configuring network policy with OpenShift SDN
```
Comment (bergerhoffer) [inline networking/configuring-networkpolicy.adoc]: Should "OpenShift SDN" be "OpenShiftSDN"?

File: modules/nw-networkpolicy-about.adoc [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+[NOTE]
+====
+The `v1` NetworkPolicy features are available only in {product-title}. This
+means that egress policy types, IPBlock, and combining `podSelector` and
```
Comment (squeed) [inline modules/nw-networkpolicy-about.adoc]: indeed, 4.1 still doesn't have support for egress policu.

**Comment by jboxman** [comment]
Thanks @bergerhoffer!

File: modules/nw-networkpolicy-about.adoc [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+In a cluster network isolation is controlled entirely by
+link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/network-policy.md[`NetworkPolicy`
+objects]. By default, all pods in a project are accessible from other pods and
+network endpoints. To isolate one or more pods in a project, you can create
+`NetworkPolicy` objects in that project to indicate the allowed incoming
+connections. Project administrators can create and delete `NetworkPolicy`
+objects within their own project.
+
+Pods that do not have `NetworkPolicy` objects pointing to them are fully
```
Comment (squeed) [inline modules/nw-networkpolicy-about.adoc]: I would replace "pointing to" with something like "selecting".

File: modules/nw-networkpolicy-about.adoc [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+  podSelector:
+    matchLabels:
+      role: frontend
+  ingress:
+  - ports:
+    - protocol: TCP
+      port: 80
+    - protocol: TCP
+      port: 443
+----
```
Comment (squeed) [inline modules/nw-networkpolicy-about.adoc]: Definitely add a network policy that shows how to grant access to the OpenShift router.

File: modules/nw-networkpolicy-create.adoc [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+[id="nw-networkpolicy-create-{context}"]
+
+= Creating a `NetworkPolicy` object
+
+To define granular rules describing ingress network traffic allowed for projects
+in your cluster, you can create `NetworkPolicy` objects.
+
+.Prerequisites
+
+* A cluster using the OpenShiftSDN network plug-in with `mode: NetworkPolicy`
```
Comment (squeed) [inline modules/nw-networkpolicy-create.adoc]: Might want to call out that this is the default.

File: networking/configuring-networkpolicy.adoc (hunk 14 above)
Comment (jboxman) [reply networking/configuring-networkpolicy.adoc]: I've been pondering this myself. We have two related, but distinct items:

- `OpenShiftSDN` is the actual Kubernetes CNI plug-in
- OpenShift SDN is our software defined network, implemented by `OpenShiftSDN` plug-in

In 3.x, it was 3 distinct plug-ins, and they had names like `ovn-multitenant`, so it was easy to avoid conflating the terms.

File: modules/nw-networkpolicy-about.adoc (hunk 17 above)
Comment (jboxman) [reply modules/nw-networkpolicy-about.adoc]: @squeed, is that this or a different policy object?

```yaml
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    name: allow-from-openshift-ingress-namespace
  spec:
    ingress:
    - from:
      - namespaceSelector:
          matchLabels:
            name: openshift-ingress
    podSelector: {}
    policyTypes:
    - Ingress
```

File: networking/configuring-networkpolicy.adoc (hunk 14 above)
Comment (danwinship) [reply networking/configuring-networkpolicy.adoc]: Hm. I would say that "`OpenShiftSDN`" is just a value that is used in certain config files, and "OpenShift SDN" is the actual name, for essentially all documentation purposes, because the documentation doesn't really need to distinguish "OpenShift SDN as our software defined network" from "OpenShift SDN as an actual Kubernetes CNI plug-in".

(It might be correct to say that "`openshift-sdn`" is the actual Kubernetes CNI plug-in, but I don't think that name ever gets used in the docs, and I don't think it needs to be.)

File: modules/nw-networkpolicy-about.adoc (hunk 3 above)
Comment (jboxman) [reply modules/nw-networkpolicy-about.adoc]: Oops, thanks!

File: modules/nw-networkpolicy-about.adoc (hunk 16 above)
Comment (jboxman) [reply modules/nw-networkpolicy-about.adoc]: Perhaps this instead?

> If a Pod is selected by one or more `NetworkPolicy` objects, then the Pod will
> accept only connections that are allowed by at least one of those `NetworkPolicy`
> objects. A Pod that is not selected by any `NetworkPolicy` objects is fully
> accessible.

File: modules/nw-networkpolicy-about.adoc [hunk 19]
```diff
@@ -6,8 +6,8 @@
 
 = About network policy
 
-In {product-title} {product-version} `NetworkPolicy` is the default mode for the
-`OpenShiftSDN` plug-in for the Kubernetes Container Network Interface (CNI).
+In {product-title} {product-version}, `NetworkPolicy` is the default mode for
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: default mode - what is a mode? What dos network policy do? Why do we care about this?

File: modules/nw-networkpolicy-about.adoc [hunk 20]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
 `namespaceSelector` are not available in {product-title}.
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: Where does this fit in? What are these things?

File: modules/nw-networkpolicy-about.adoc [hunk 21]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
 ====
 
-In a cluster network isolation is controlled entirely by
+In a cluster, network isolation is controlled entirely by
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: In a cluster using network policy, ??

File: modules/nw-networkpolicy-about.adoc [hunk 22]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
 link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/network-policy.md[`NetworkPolicy`
-objects]. By default, all pods in a project are accessible from other pods and
-network endpoints. To isolate one or more pods in a project, you can create
+objects]. By default, all Pods in a project are accessible from other Pods and
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: ..other Podes and network endpoints --- in what? the cluster, namespace, internet?

File: modules/nw-networkpolicy-about.adoc [hunk 23]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
+network endpoints. To isolate one or more Pods in a project, you can create
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: isolate from what?

File: modules/nw-networkpolicy-about.adoc [hunk 24]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
 `NetworkPolicy` objects in that project to indicate the allowed incoming
 connections. Project administrators can create and delete `NetworkPolicy`
 objects within their own project.
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: Is ownership a factor here? Why do we mention Project admins? Are cluster admins handled differently?

File: modules/nw-networkpolicy-about.adoc [hunk 25]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
 
-Pods that do not have `NetworkPolicy` objects pointing to them are fully
-accessible, whereas, pods that have one or more `NetworkPolicy` objects pointing
-to them are isolated. These isolated pods only accept connections that are
-accepted by at least one of their `NetworkPolicy` objects.
+If a Pod is selected by one or more `NetworkPolicy` objects, then the Pod will
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: What does selected by a network policy object mean?

File: modules/nw-networkpolicy-about.adoc [hunk 26]
```diff
@@ -16,26 +16,26 @@ means that egress policy types, IPBlock, and combining `podSelector` and
...
+accept only connections that are allowed by at least one of those `NetworkPolicy`
+objects. A Pod that is not selected by any `NetworkPolicy` objects is fully
+accessible.
 
-Following are a few sample `NetworkPolicy` object definitions supporting
-different scenarios:
+The following example `NetworkPolicy` objects demonstrate supporting different
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: In the

File: modules/nw-networkpolicy-about.adoc [hunk 27]
```diff
@@ -48,10 +48,33 @@ spec:
   ingress: []
 ----
 
-* *Only Accept connections from pods within project*
+* Only allow connections from the {product-title} ingress router:
 +
-To make pods accept connections from other pods in the same project,
-but reject all other connections from pods in other projects:
+* To make a project allow only connections from the {product-title} ingress
+router, add the following `NetworkPolicy` object:
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: below it is ingress namespace

File: modules/nw-networkpolicy-about.adoc [hunk 28]
```diff
@@ -94,7 +117,7 @@ spec:
 
 For example, for the `NetworkPolicy` objects defined in previous samples, you
 can define both `allow-same-namespace` and `allow-http-and-https` policies
-within the same project. Thus allowing the pods with the label `role=frontend`,
+within the same project. Thus allowing the Pods with the label `role=frontend`,
```
Comment (pecameron) [inline modules/nw-networkpolicy-about.adoc]: "Thus allowing the Pods with the label `role=frontend`,"   This clears up a lot. It should appear sooner.

**Review by pecameron — COMMENTED** [review]
sorry, I got picky
I am not all that familiar with network policy. So I was reading to get some understanding. I need to read the section that lays out the problem and our approach to solving it. Diving in here doesn't help me understand network policy. There is a lot new jargon and concepts that relate to each other but I am not sure how it all fits together.

File: modules/nw-networkpolicy-about.adoc (hunk 17 above)
Comment (squeed) [reply modules/nw-networkpolicy-about.adoc]: That is right... but I have to get a PR in to the ingress operator to make it actually work! Good think we thought about this.

Comment (jboxman) [reply modules/nw-networkpolicy-about.adoc]: @squeed, I'll comment out the part that references this until it's merged. Thanks!

Comment (squeed) [reply modules/nw-networkpolicy-about.adoc]: Waiting for https://github.com/openshift/cluster-ingress-operator/pull/218 to merge.

**Comment by jboxman** [comment]
@pecameron, thanks for the comments!

File: modules/nw-networkpolicy-about.adoc (hunk 27 above)
Comment (jboxman) [reply modules/nw-networkpolicy-about.adoc]: @pecameron, does the ingress router live in the ingress namespace? This policy object was provided in OSDOCS-374 by @zhaozhanqi.

**Comment by jboxman** [comment]
@vikram-redhat, merged.
